macros/calendar_date/string_to_date.sql

```diff
@@ -0,0 +1,152 @@
+{%- macro string_to_date(date_string, format='yyyy-mm-dd') -%}
+    {{ adapter.dispatch('string_to_date', 'dbt_date') (date_string, format) }}
+{%- endmacro -%}
+
+{%- macro default__string_to_date(date_string, format) -%}
+-- redshift and postgres are actually pretty good at detecting date strings. we'll add compatibility for Day/Month/Year format
+
+{% if format|lower in ('d/m/yy', 'dd/mm/yy') %}
+-- Ex: (0)1/(4)/22
+to_date('{{ date_string }}', 'dd/mm/yy')
+
+{% elif format|lower in ('d/m/yyyy', 'dd/mm/yyyy') %}
+-- Ex: (0)1/(4)/2022
+to_date('{{ date_string }}', 'dd/mm/yyyy')
+
+{% else %}
+cast('{{ date_string }}' as date)
+
+{% endif %}
+
+{%- endmacro -%}
+
+{%- macro bigquery__string_to_date(date_string, format) -%}
+
+{% if format|lower[:3] = 'mon' %}
+-- Ex: Apr(il) (0)1, 2022
+parse_date("%b %e, %Y", '{{ date_string }}')
+
+{% elif format|lower[:5] = 'd mon' or format|lower[:6] = 'dd mon' %}
+-- Ex: (0)1 Apr(il) 2022
+parse_date("%e %b %Y", '{{ date_string }}')
```

Comment:
for all of these calls to `parse_date` and their equivalents on the other platforms, I'd love it if we refactored that a bit so we only do that once at the end after building the correct format string for the particular case. So, e.g. this pseudocode
```
if format = "x"
    sql_format = "y"
elseif
   ...
else
    sql_format = "z"

parse_date(sql_format, date)
```

Does that make sense? So for each platform a section where we evaluate the passed in `format` to build the `sql_format`, then pass that to the respective date function for the platform.